mage_ai/cluster_manager/manage.py

```diff
@@ -0,0 +1,304 @@
+                    workspace['project_uuid'] = project_uuid
+                    workspace['config'] = config.get('config')
+                    if user:
+                        workspace['access'] = user.get_access(
+                            Entity.PROJECT,
+                            project_uuid,
+                        )
+                workspaces.append(workspace)
+
+            except Exception as e:
+                print(f'Error fetching workspace: {str(e)}')
+    return workspaces
+
+
+def check_auto_termination(cluster_type: ClusterType):
+    if cluster_type == ClusterType.K8S:
+        from mage_ai.cluster_manager.kubernetes.workload_manager import WorkloadManager
+
+        namespace = os.getenv(KUBE_NAMESPACE, 'default')
+        workload_manager = WorkloadManager(namespace=namespace)
+
+        workspaces = get_workspaces(cluster_type)
+        for ws in workspaces:
+            workspace_config = WorkspaceConfig.load(ws.get('config', {}))
+            termination_policy = workspace_config.termination_policy
+            if termination_policy.enable_auto_termination:
+                activity_details = workload_manager.get_workload_activity(
+                    ws.get('name')
+                )
+                max_idle_seconds = termination_policy.max_idle_seconds
+                if activity_details and max_idle_seconds > 0:
+                    active_pipeline_run_count = activity_details.get(
+                        'active_pipeline_run_count'
+                    )
+
+                    last_user_request_ts = activity_details.get('last_user_request')
+                    last_user_request = datetime.fromisoformat(last_user_request_ts)
+
+                    latest_activity_time = last_user_request.timestamp()
+                    last_scheduler_activity_ts = activity_details.get(
+                        'last_scheduler_activity'
+                    )
+                    if last_scheduler_activity_ts:
+                        last_scheduler_activity = datetime.fromisoformat(
+                            last_scheduler_activity_ts
+                        )
+                        latest_activity_time = max(
+                            latest_activity_time,
+                            last_scheduler_activity.timestamp(),
+                        )
+
+                    now_time = datetime.utcnow().timestamp()
+                    if (
+                        not active_pipeline_run_count
+                        and now_time - latest_activity_time > max_idle_seconds
+                    ):
+                        workload_manager.scale_down_workload(ws.get('name'))
+
+
+def create_workspace(
+    cluster_type: ClusterType,
+    workspace_name: str,
+    workspace_config: WorkspaceConfig,
+    payload: Dict,
+):
+    workspace_file = None
+    project_uuid = None
+    project_type = get_project_type()
+    if project_type == ProjectType.MAIN:
+        project_folder = os.path.join(get_repo_path(), 'projects')
```

Comment:
I saw it in a few places. I would suggest creating a method for getting the projects folder.
I would also suggest creating a class for Workspace.
In the Workspace class and put the workspace CRUD methods in it.

Reply:
ok i'll do this in the next PR